### PR TITLE
Enhancement horizontal gap port and new method create_wave_port()

### PR DESCRIPTION
**Pull request by ring630**

### CODE REVIEW

**Comment by ring630** [comment]
Improved create_edge_port(). Now the user can specify where to place the negative port terminal.

```
def create_edge_port(
        self,
        primivitivename,
        edgenumber,
        iscircuit=False,
        iswave=False,
        wave_horizontal_extension=5,
        wave_vertical_extension=3,
        wave_launcher="1mm",
        ref_primitive_name=None,
        ref_edge_number=0,
    ):
```

File: pyaedt/hfss3dlayout.py [hunk 1]
```diff
@@ -218,6 +233,78 @@ def create_edge_port(
...
+        wave_vertical_extension=3,
+        wave_launcher="1mm",
+    ):
+        """Create a single ended wave port.
+
+        Parameters
+        ----------
+        primivitive_name : str
+            Name of the primitive to create the edge port on.
+        edge_number :
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:251]:
```suggestion
        edge_number : int
```

File: pyaedt/hfss3dlayout.py [hunk 2]
```diff
@@ -218,6 +233,78 @@ def create_edge_port(
...
+    @pyaedt_function_handler()
+    def create_wave_port(
+        self,
+        primivitive_name,
+        edge_number,
+        wave_horizontal_extension=5,
+        wave_vertical_extension=3,
+        wave_launcher="1mm",
+    ):
+        """Create a single ended wave port.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:245]:
```suggestion
        """Create a single-ended wave port.
```

File: pyaedt/hfss3dlayout.py [hunk 3]
```diff
@@ -218,6 +233,78 @@ def create_edge_port(
...
+        Parameters
+        ----------
+        primivitive_name : str
+            Name of the primitive to create the edge port on.
+        edge_number :
+            Edge number to create the edge port on.
+        wave_horizontal_extension : float, optional
+            Horizontal port extension factor. The default is `5`.
+        wave_vertical_extension : float, optional
+            Vertical port extension factor. The default is `5`.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:256]:
```suggestion
            Vertical port extension factor. The default is ``5``.
```

File: pyaedt/hfss3dlayout.py [hunk 4]
```diff
@@ -218,6 +233,78 @@ def create_edge_port(
...
+        """Create a single ended wave port.
+
+        Parameters
+        ----------
+        primivitive_name : str
+            Name of the primitive to create the edge port on.
+        edge_number :
+            Edge number to create the edge port on.
+        wave_horizontal_extension : float, optional
+            Horizontal port extension factor. The default is `5`.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:254]:
```suggestion
            Horizontal port extension factor. The default is ``5``.
```

File: pyaedt/hfss3dlayout.py [hunk 5]
```diff
@@ -218,6 +233,78 @@ def create_edge_port(
...
+        wave_vertical_extension : float, optional
+            Vertical port extension factor. The default is `5`.
+        wave_launcher : str, optional
+            PEC (perfect electrical conductor) launcher size with units. The
+            default is `"1mm"`.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:259]:
```suggestion
            default is ``"1mm"``.
```

File: pyaedt/hfss3dlayout.py [hunk 6]
```diff
@@ -160,6 +162,11 @@ def create_edge_port(
         wave_launcher : str, optional
             PEC (perfect electrical conductor) launcher size with units. The
             default is `"1mm"`.
+        ref_primitive_name : str, optional
+            Name of the reference primitive to place negative edge port terminal.
+            The default is ``None``.
+        ref_edge_number : str, int
+            Edge number of reference primitive.The default is ``0``.
```
Comment (MaxJPRey) [inline pyaedt/hfss3dlayout.py:169]:
```suggestion
            Edge number of reference primitive. The default is ``0``.
```

**Review by MaxJPRey — APPROVED** [review]
LGTM. Thanks @ring630 for adding this new way to create port.

**Pull request merged**